src/cht/map/bucket.rs

```diff
@@ -368,34 +355,63 @@ impl<'g, K: 'g + Eq, V: 'g> BucketArray<K, V> {
     }
 }
 
-impl<'g, K: 'g, V: 'g> BucketArray<K, V> {
-    fn probe_loop<F, T>(&self, guard: &'g Guard, hash: u64, mut f: F) -> ProbeLoopResult<T>
-    where
-        F: FnMut(usize, &Atomic<Bucket<K, V>>, Shared<'g, Bucket<K, V>>) -> ProbeLoopAction<T>,
-    {
-        let offset = hash as usize & (self.buckets.len() - 1);
+struct Probe<'b, 'g, K: 'g, V: 'g> {
+    buckets: &'b [Atomic<Bucket<K, V>>],
+    guard: &'g Guard,
+    this_bucket: (usize, &'b Atomic<Bucket<K, V>>),
+    offset: usize,
 
-        for i in
-            (0..self.buckets.len()).map(|i| (i.wrapping_add(offset)) & (self.buckets.len() - 1))
-        {
-            let this_bucket = &self.buckets[i];
+    i: usize,
+    reload: bool,
+}
 
-            loop {
-                let this_bucket_ptr = this_bucket.load_consume(guard);
+impl<'b, 'g, K: 'g, V: 'g> Probe<'b, 'g, K, V> {
+    fn reload(&mut self) {
+        self.reload = true;
+    }
+}
 
-                if is_sentinel(this_bucket_ptr) {
-                    return ProbeLoopResult::FoundSentinelTag;
-                }
+impl<'b, 'g, K: 'g, V: 'g> Iterator for Probe<'b, 'g, K, V> {
+    type Item = Result<(usize, &'b Atomic<Bucket<K, V>>, Shared<'g, Bucket<K, V>>), ()>;
 
-                match f(i, this_bucket, this_bucket_ptr) {
-                    ProbeLoopAction::Continue => break,
-                    ProbeLoopAction::Reload => (),
-                    ProbeLoopAction::Return(t) => return ProbeLoopResult::Returned(t),
-                }
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.reload {
+            let max = self.buckets.len() - 1;
+            if self.i >= max {
+                return None;
             }
+            self.i += 1;
+            let i = self.i.wrapping_add(self.offset) & max;
+            self.this_bucket = (i, &self.buckets[i]);
         }
+        self.reload = false;
+
+        let this_bucket_ptr = self.this_bucket.1.load_consume(self.guard);
 
-        ProbeLoopResult::LoopEnded
+        if is_sentinel(this_bucket_ptr) {
+            return Some(Err(()));
+        }
+
+        let val = (self.this_bucket.0, self.this_bucket.1, this_bucket_ptr);
+        Some(Ok(val))
+    }
+}
+
+impl<'g, K: 'g, V: 'g> BucketArray<K, V> {
+    fn probe(&self, guard: &'g Guard, hash: u64) -> Probe<'_, 'g, K, V> {
+        let buckets = &self.buckets;
+        let offset = hash as usize & (buckets.len() - 1);
+        // FIXME: this will panic if `len() == 0`
+        let this_bucket = (offset, &buckets[offset]);
```

Comment:
This is a change I’m a bit uncertain about, maybe using an `Option` for `this_bucket` could avoid that, and also remove some panic code as well. If you mention you want to optimize the probing itself either way, it might not be worth worth doing this immediately, if indeed `len` will never be 0.

Reply:
I reviewed the `cht` code and confirmed that `len` will never be 0. So this line will never cause a panic.

`BucketArray` can be created by `default` method or `with_length` function.

`default` uses 128 as `len`:

https://github.com/moka-rs/moka/blob/e5690f8932b5d22ea65cd5d11584ccaf4a72049e/src/cht/map/bucket.rs#L28

`with_length` takes an arbitrary `len` in `usize`, but it's caller (below) will never call it with 0:

https://github.com/moka-rs/moka/blob/e5690f8932b5d22ea65cd5d11584ccaf4a72049e/src/cht/segment.rs#L170-L184

| `capacity` | `actual_num_segments` | `actual_capacity` (`len`) |
|:--:|:--:|:--:|
| 0 | any | n/a (does not call `with_length`) |
| 1 | 1 | 2 |
| 1 | 2 | 1 |
| 2 | 1 | 4 |